### PR TITLE
ENH: Update eigen3 to top of upstream branch 3.3

### PR DESCRIPTION
**Pull request by phcerdan**
Top is commit: [1553e2](https://bitbucket.org/eigen/eigen/commits/1553e26d9ce848a947e25225fec6fb2c58c2b921) in branch [3.3](https://bitbucket.org/eigen/eigen/src/3.3/)

Also rebase all Kitware changes on top of eigen to facilitate updates

Closes https://github.com/InsightSoftwareConsortium/ITK/pull/1290

### CODE REVIEW

**Comment by thewtex** [comment]
@phcerdan thank you!

A few Windows build errors.

**Comment by phcerdan** [comment]
Not sure yet where the error is actually triggered with that template explosion.

It seems that somewhere we are using a float instead of an integer type (only in windows?)
Maybe useful: http://eigen.tuxfamily.org/index.php?title=3.3#Index_typedef

**Comment by phcerdan** [comment]
I cannot reproduce using local windows build with `Debug` and tests enabled.
local visual studio version is: 15.9.16

Maybe is worth to report upstream, to check if there are some known problems with the compiler used in our azure pipelines?

**Comment by dzenanz** [comment]
Instantiating `SymmetricSecondRankTensor` with type double triggers that warning that integer-like type was expected.
https://github.com/InsightSoftwareConsortium/ITK/blob/5ca7d26ecf283a1586e3ade04e8da77bd370d438/Modules/Core/Common/test/itkSymmetricSecondRankTensorTest.cxx#L200-L202

I can reproduce it with VS2019 (version 16.3.3). But I don't know how to fix it.

**Comment by thewtex** [comment]
> Instantiating `SymmetricSecondRankTensor` with type **double** triggers the warning that integer-like type was expected.

Is there any more information on the line number? e.g. traceback?

**Comment by dzenanz** [comment]
```text
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/PlainObjectBase.h(742,1): error C2338: FLOATING_POINT_ARGUMENT_PASSED__INTEGER_WAS_EXPECTED
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message : see reference to function template instantiation 'void itkeigen::PlainObjectBase<itkeigen::Matrix<double,3,3,0,3,3>>::_init2<T0,T1>(itkeigen::EigenBase<Derived>::Index,itkeigen::EigenBase<Derived>::Index,__int64 *)' being compiled
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :         with
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :         [
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :             T0=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :             T1=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :             Derived=itkeigen::Matrix<double,3,3,0,3,3>
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :         ]
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message : see reference to function template instantiation 'void itkeigen::PlainObjectBase<itkeigen::Matrix<double,3,3,0,3,3>>::_init2<T0,T1>(itkeigen::EigenBase<Derived>::Index,itkeigen::EigenBase<Derived>::Index,__int64 *)' being compiled
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :         with
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :         [
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :             T0=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :             T1=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :             Derived=itkeigen::Matrix<double,3,3,0,3,3>
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Core/Matrix.h(302): message :         ]
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(167): message : see reference to function template instantiation 'itkeigen::Matrix<double,3,3,0,3,3>::Matrix<itkeigen::EigenBase<Derived>::Index,itkeigen::EigenBase<Derived>::Index>(const T0 &,const T1 &)' being compiled
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(167): message :         with
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(167): message :         [
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(167): message :             Derived=itkeigen::Matrix<ValueType,3,3,1,3,3>,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(167): message :             T0=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(167): message :             T1=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(167): message :         ]
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(163): message : see reference to function template instantiation 'itkeigen::Matrix<double,3,3,0,3,3>::Matrix<itkeigen::EigenBase<Derived>::Index,itkeigen::EigenBase<Derived>::Index>(const T0 &,const T1 &)' being compiled
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(163): message :         with
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(163): message :         [
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(163): message :             Derived=itkeigen::Matrix<ValueType,3,3,1,3,3>,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(163): message :             T0=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index,
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(163): message :             T1=itkeigen::EigenBase<itkeigen::Matrix<ValueType,3,3,1,3,3>>::Index
4>C:\Dev\ITK-git\Modules\ThirdParty\Eigen3\src\itkeigen\Eigen\src/Eigenvalues/SelfAdjointEigenSolver.h(163): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message : see reference to function template instantiation 'itkeigen::SelfAdjointEigenSolver<EigenLibMatrixType>::SelfAdjointEigenSolver<Derived>(const itkeigen::EigenBase<Derived> &,int)' being compiled
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :             Derived=itkeigen::Matrix<ValueType,3,3,1,3,3>
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message : see reference to function template instantiation 'itkeigen::SelfAdjointEigenSolver<EigenLibMatrixType>::SelfAdjointEigenSolver<Derived>(const itkeigen::EigenBase<Derived> &,int)' being compiled
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :             Derived=itkeigen::Matrix<ValueType,3,3,1,3,3>
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(951): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message : see reference to function template instantiation 'unsigned int itk::SymmetricEigenAnalysisFixedDimension<3,itk::Matrix<itk::SpacePrecisionType,3,3>,itk::FixedArray<T,3>,itk::Matrix<T,3,3>>::ComputeEigenValuesAndVectorsWithEigenLibraryImpl<TMatrix>(const QMatrix &,TVector &,TEigenMatrix &,long) const' being compiled
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             T=double,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             TMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             QMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             TVector=itk::FixedArray<double,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             TEigenMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message : see reference to function template instantiation 'unsigned int itk::SymmetricEigenAnalysisFixedDimension<3,itk::Matrix<itk::SpacePrecisionType,3,3>,itk::FixedArray<T,3>,itk::Matrix<T,3,3>>::ComputeEigenValuesAndVectorsWithEigenLibraryImpl<TMatrix>(const QMatrix &,TVector &,TEigenMatrix &,long) const' being compiled
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             T=double,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             TMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             QMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             TVector=itk::FixedArray<double,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :             TEigenMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(789): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message : while compiling class template member function 'unsigned int itk::SymmetricEigenAnalysisFixedDimension<3,itk::Matrix<itk::SpacePrecisionType,3,3>,itk::FixedArray<T,3>,itk::Matrix<T,3,3>>::ComputeEigenValuesAndVectors(const TMatrix &,TVector &,TEigenMatrix &) const'
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message :             T=double,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message :             TMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message :             TVector=itk::FixedArray<double,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message :             TEigenMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricEigenAnalysis.h(788): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message : see reference to function template instantiation 'unsigned int itk::SymmetricEigenAnalysisFixedDimension<3,itk::Matrix<itk::SpacePrecisionType,3,3>,itk::FixedArray<T,3>,itk::Matrix<T,3,3>>::ComputeEigenValuesAndVectors(const TMatrix &,TVector &,TEigenMatrix &) const' being compiled
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message :             T=double,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message :             TMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message :             TVector=itk::FixedArray<double,3>,
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message :             TEigenMatrix=itk::Matrix<itk::SpacePrecisionType,3,3>
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(295): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(259): message : see reference to class template instantiation 'itk::SymmetricEigenAnalysisFixedDimension<3,itk::Matrix<itk::SpacePrecisionType,3,3>,itk::FixedArray<T,3>,itk::Matrix<T,3,3>>' being compiled
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(259): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(259): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(259): message :             T=double
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(259): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(258): message : while compiling class template member function 'void itk::SymmetricSecondRankTensor<double,3>::ComputeEigenValues(itk::FixedArray<T,3> &) const'
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(258): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(258): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(258): message :             T=double
4>C:\Dev\ITK-git\Modules\Core\Common\include\itkSymmetricSecondRankTensor.hxx(258): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\test\itkSymmetricSecondRankTensorTest.cxx(227): message : see reference to function template instantiation 'void itk::SymmetricSecondRankTensor<double,3>::ComputeEigenValues(itk::FixedArray<T,3> &) const' being compiled
4>C:\Dev\ITK-git\Modules\Core\Common\test\itkSymmetricSecondRankTensorTest.cxx(227): message :         with
4>C:\Dev\ITK-git\Modules\Core\Common\test\itkSymmetricSecondRankTensorTest.cxx(227): message :         [
4>C:\Dev\ITK-git\Modules\Core\Common\test\itkSymmetricSecondRankTensorTest.cxx(227): message :             T=double
4>C:\Dev\ITK-git\Modules\Core\Common\test\itkSymmetricSecondRankTensorTest.cxx(227): message :         ]
4>C:\Dev\ITK-git\Modules\Core\Common\test\itkSymmetricSecondRankTensorTest.cxx(202): message : see reference to class template instantiation 'itk::SymmetricSecondRankTensor<double,3>' being compiled
```

**Comment by hjmjohnson** [comment]
@phcerdan FYI:  I am running into Eigen issues in ITK, and I think this PR will fix some of them.  Do you have an estimate of what needs to be done to complete this issue?

**Comment by phcerdan** [comment]
@hjmjohnson, I cannot reproduce the issue in my local windows, so no time estimate, I am sorry. I do not understand how the problem is only triggered in a recent version of `msvc`, and only after a few new commits on the Eigen side. I am thinking this might be in the compiler side... What kind of problems are you getting with the current internal Eigen used in ITK?

**Comment by hjmjohnson** [comment]
> What kind of problems are you getting with the current internal Eigen used in ITK?

@phcerdan 
```
/home/johnsonhj/Dashboard/src/ITK/Modules/ThirdParty/Eigen3/src/itkeigen/Eigen/src/Core/arch/AVX/PacketMath.h:169:54: error: invalid output size for constraint '+x'
  __asm__("vfmadd231ps %[a], %[b], %[c]" : [c] "+x" (res) : [a] "x" (a), [b] "x" (b));
                                                     ^
/home/johnsonhj/Dashboard/src/ITK/Modules/ThirdParty/Eigen3/src/itkeigen/Eigen/src/Core/arch/AVX/PacketMath.h:179:54: error: invalid output size for constraint '+x'
  __asm__("vfmadd231pd %[a], %[b], %[c]" : [c] "+x" (res) : [a] "x" (a), [b] "x" (b));
                                                     ^
```

I am pretty sure that this problem as been addressed in upstream versions of Eigen.  I get these errors for both mac and ubuntu builds.

**Comment by phcerdan** [comment]
I have reported an issue in Eigen3, asking for any ideas/hints about this. http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1761

**Comment by hjmjohnson** [comment]
@phcerdan I'm still reviewing. I ran into a bit of a red-herring in that I compiled upstream with "-mtune=native -march=i7" and I compiled the master branch with "-mtune=native -march=native".  It seems that the problem only occurs when optimization is set ot "-mtune=native -march=native".

I"m looking more closely at this now.

**Comment by phcerdan** [comment]
Well, I found the buggy line in Eigen, http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1761#c12 not sure yet if the problem is in Eigen or in the compiler (msvc2019), but I will let anyone with more knowledge than me with msvc internal API to report that problem upstream, or let the Eigen developers solve the issue internally.

When this is done, we would be able to update and merge this PR.

**Comment by thewtex** [comment]
Awesome work @phcerdan ! :1st_place_medal: 

@dzenanz could you please submit a bug report the MSVC team?

**Comment by phcerdan** [comment]
I am leaning on the bug being in the Eigen side though..., I am reporting back there after a few questions of Eigen developer.

**Comment by phcerdan** [comment]
Solved upstream, updating it to the new top of the branches/3.3

**Comment by dzenanz** [comment]
> @dzenanz could you please submit a bug report the MSVC team?

I looked at the discussion on Eigen bug tracker, and concluded to wait for them to resolve or submit a bug to Microsoft.

File: Modules/ThirdParty/Eigen3/src/itkeigen/Eigen/Eigenvalues [hunk 1]
```diff
@@ -10,14 +10,14 @@
...
 
-#include "src/Core/util/DisableStupidWarnings.h"
-
 #include "Cholesky"
 #include "Jacobi"
 #include "Householder"
 #include "LU"
 #include "Geometry"
 
+#include "src/Core/util/DisableStupidWarnings.h"
```
Comment (thewtex) [inline Modules/ThirdParty/Eigen3/src/itkeigen/Eigen/Eigenvalues]: :laughing:

**Comment by phcerdan** [comment]
All green, merging.

**Pull request merged**